src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUMemoryManager.java

```diff
@@ -518,21 +519,28 @@ public void clearMemory() {
 	/**
 	 * Clears up the memory used by non-dirty pointers.
 	 */
-	public void clearTemporaryMemory() {
-		// To record the cuda block sizes needed by allocatedGPUObjects, others are cleared up.
-		Set<Pointer> unlockedDirtyPointers = matrixMemoryManager.getPointers(false, true, false);
-		Set<Pointer> temporaryPointers = nonIn(allPointers.keySet(), unlockedDirtyPointers);
+	public void clearTemporaryMemory(HashSet<MatrixObject> outputMatrixObjects) {
+		Set<Pointer> donotClearPointers =  new HashSet<>();
+		// First clean up all GPU objects except:
+		// 1. Output matrix objects
+		// 2. GPU objects that are currently being used (i.e. locked)
+		// 3. Matrix object are 
```

Comment:
Just for the sake of clarity might want to comment that we also do not clean up pinned inputs